src/charm.py

```diff
@@ -0,0 +1,222 @@
+from managers.tls import TLSManager
+from workload import ODWorkload
+
+# from events.upgrade import ODUpgradeEvents, OpensearchDasboardsDependencyModel
+
+
+logger = logging.getLogger(__name__)
+
+
+class OpensearchDasboardsCharm(CharmBase):
+    """Charmed Operator for Opensearch Dashboards."""
+
+    def __init__(self, *args):
+        super().__init__(*args)
+        self.name = CHARM_KEY
+        self.state = ClusterState(self, substrate=SUBSTRATE)
+        self.workload = ODWorkload()
+
+        # --- CHARM EVENT HANDLERS ---
+
+        self.password_action_events = PasswordActionEvents(self)
+        self.tls_events = TLSEvents(self)
+        self.requirer_events = RequirerEvents(self)
+        # self.upgrade_events = ODUpgradeEvents(
+        #     self,
+        #     dependency_model=OpensearchDasboardsDependencyModel(
+        #         **DEPENDENCIES  # pyright: ignore[reportGeneralTypeIssues]
+        #     ),
+        # )
+
+        # --- MANAGERS ---
+
+        self.tls_manager = TLSManager(
+            state=self.state, workload=self.workload, substrate=SUBSTRATE
+        )
+        self.config_manager = ConfigManager(
+            state=self.state, workload=self.workload, substrate=SUBSTRATE, config=self.config
+        )
+
+        # --- LIB EVENT HANDLERS ---
+
+        self.restart = RollingOpsManager(self, relation="restart", callback=self._restart)
+
+        # --- CORE EVENTS ---
+
+        self.framework.observe(getattr(self.on, "install"), self._on_install)
+
+        self.framework.observe(getattr(self.on, "start"), self._start)
+
+        self.framework.observe(
+            getattr(self.on, "update_status"), self._on_cluster_relation_changed
+        )
+        self.framework.observe(
+            getattr(self.on, "leader_elected"), self._on_cluster_relation_changed
+        )
+        self.framework.observe(
+            getattr(self.on, "config_changed"), self._on_cluster_relation_changed
+        )
+
+        self.framework.observe(
+            getattr(self.on, f"{PEER}_relation_changed"), self._on_cluster_relation_changed
+        )
+        self.framework.observe(
+            getattr(self.on, f"{PEER}_relation_joined"), self._on_cluster_relation_changed
+        )
+        self.framework.observe(
+            getattr(self.on, f"{PEER}_relation_departed"), self._on_cluster_relation_changed
+        )
+
+        self.framework.observe(getattr(self.on, "secret_changed"), self._on_secret_changed)
+
+    # --- CORE EVENT HANDLERS ---
+
+    def _on_install(self, event: InstallEvent) -> None:
+        """Handler for the `on_install` event."""
+        self.unit.status = MaintenanceStatus("installing Opensearch Dashboards...")
+
+        install = self.workload.install()
+        if not install:
+            self.unit.status = BlockedStatus("unable to install Opensearch Dashboards")
+
+        # don't complete install until passwords set
+        if not self.state.peer_relation:
```

Comment:
I believe this check will resolve to `False` if we have a single unit (as it boils down to: `self.model.get_relation(PEER)`). The alternative would be to check if the peer_relation exists, then get the password.

Reply:
Hmmm, we DO have a relation when it's a single unit...

(I may have got something wrong here?)

Reply:
@juditnovak just tested, indeed it renders a relation even with a single unit. Test on Juju 3.4.2.